array/sort.mbt

```diff
@@ -0,0 +1,394 @@
+// Copyright 2024 International Digital Economy Academy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+struct ArraySlice[T] {
+  array : Array[T]
+  start : Int
+  end : Int
+}
+
+fn length[T](self : ArraySlice[T]) -> Int {
+  self.end - self.start
+}
+
+fn op_get[T](self : ArraySlice[T], index : Int) -> T {
+  self.array[self.start + index]
+}
+
+fn op_set[T](self : ArraySlice[T], index : Int, value : T) -> Unit {
+  self.array[self.start + index] = value
+}
+
+fn swap[T](self : ArraySlice[T], a : Int, b : Int) -> Unit {
+  self.array.swap(self.start + a, self.start + b)
+}
+
+fn reverse[T](self : ArraySlice[T]) -> Unit {
+  let mid_len = self.length() / 2
+  for i = 0; i < mid_len; i = i + 1 {
+    let j = self.length() - i - 1
+    self.swap(i, j)
+  }
+}
+
+fn slice[T](self : ArraySlice[T], start : Int, end : Int) -> ArraySlice[T] {
+  { array: self.array, start: self.start + start, end: self.start + end }
+}
+
+/// Sorts the array
+/// 
+/// It's an in-place, unstable sort(it will reorder equal elements). The time complexity is O(n log n) in the worst case.
+/// 
+/// # Example
+/// 
+/// ```
+/// let arr = [5, 4, 3, 2, 1]
+/// arr.sort()
+/// debug(arr) //output: [1, 2, 3, 4, 5]
+/// ```
+pub fn sort[T : Compare](self : Array[T]) -> Unit {
+  quick_sort(
+    { array: self, start: 0, end: self.length() },
+    T::compare,
```

Comment:
这种额外开销很高吗？还有什么写法能够比较通用的？如果几种不同的 sort 方式如果都要硬编码 inline 的话有点太丑了

Reply:
The overhead comes from passing the function `T::compare`, so I'd recommend you just stick with generic functions, ie. using `[T: Compare]`. The moonbit compiler implements monomorphization so the runtime cost of generic functions is low.

Reply:
But it means we cannot reuse the same code for `sort_by` and `sort_by_key`, right?

Reply:
I'll do a benchmark to see the performance impact

Reply:
It's about 20% faster when using [T: Compare] directly, tested on my machine (Mac 2020 M1), with 1M random values

Reply:
Is it possible for Moonbit to have specializations of high-order functions in the future?

Reply:
We need to discuss about it in our team. /cc @bobzhang

Reply:
We will perform more advanced optimizations in the future, however, it is recommended to write obviously fast code for perf critical code instead of relying on black box advanced optimizations, you may encounter perf cliff from time to time.